src/source/load_vector_tile.ts

```diff
@@ -38,46 +62,143 @@ export class DedupedRequest {
+        err: Error | null | undefined;
+        result: any;
+    }) {
+        if (this.scheduler) {
+            this.scheduler.add(() => {
+                callback(err, result);
+            }, metadata);
+        } else {
+            callback(err, result);
+        }
+    }
+
+    getEntry = (key: string) => {
+        return (
+            this.entries[key] || {
+                // use a set to avoid duplicate callbacks being added when calling from queue
+                callbacks: new Set(),
+            }
+        );
+    };
+
+    request({key, metadata, requestFunc, callback, fromQueue}: DedupedRequestInput): Cancelable {
+        const entry = (this.entries[key] = this.getEntry(key));
+
+        const removeCallbackFromEntry = ({key, requestCallback}) => {
+            const entry = this.getEntry(key);
+            if (entry.result) return;
+            entry.callbacks.delete(requestCallback);
+            if (entry.callbacks.size) {
+                return;
+            }
+            if (entry.cancel) {
+                entry.cancel();
+            }
+            filterQueue(key);
+            delete this.entries[key];
+        };
+
+        let advanced = false;
+        const advanceImageRequestQueue = () => {
+            if (advanced) {
+                return;
+            }
+            advanced = true;
+            numRequests--;
+            assert(numRequests >= 0);
+            while (requestQueue.size && numRequests < 50) {
+                const request = requestQueue.values().next().value;
+                const {key, metadata, requestFunc, callback, cancelled} = request;
+                filterQueue(key);
+                if (!cancelled) {
+                    request.cancel = this.request({
+                        key,
+                        metadata,
+                        requestFunc,
+                        callback,
+                        fromQueue: true
+                    }).cancel;
+                }
+            }
+        };
 
         if (entry.result) {
             const [err, result] = entry.result;
-            if (this.scheduler) {
-                this.scheduler.add(() => {
-                    callback(err, result);
-                }, metadata);
-            } else {
-                callback(err, result);
-            }
-            return () => {};
+            this.addToSchedulerOrCallDirectly({
+                callback,
+                metadata,
+                err,
+                result,
+            });
+            return {cancel: () => {}};
         }
 
-        entry.callbacks.push(callback);
+        entry.callbacks.add(callback);
 
-        if (!entry.cancel) {
-            entry.cancel = request((err, result) => {
+        const inQueue = requestQueue.has(key);
+        if ((!entry.cancel && !inQueue) || fromQueue) {
+            // Lack of attached cancel handler means this is the first request for this resource
+            if (numRequests >= 50) {
```

Comment:
When we're receiving too many requests, shove stuff on the queue, and return the queued object which has a cancellation handler on it, so that it can be cancelled by the caller